evalml/pipelines/components/transformers/samplers/base_sampler.py

```diff
@@ -54,3 +58,67 @@ def transform(self, X, y=None):
         if y is not None:
             y = infer_feature_types(y)
         return X, y
+
+
+class BaseOverSampler(BaseSampler):
+    """Base Oversampler component. Used as the base class of all imbalance-learn oversampler components"""
+
+    def __init__(self, sampler, parameters, component_obj, random_seed):
+        error_msg = "imbalanced-learn is not installed. Please install using 'pip install imbalanced-learn'"
+        im = import_or_raise("imblearn.over_sampling", error_msg=error_msg)
+        self.sampler = {"SMOTE": im.SMOTE,
+                        "SMOTENC": im.SMOTENC,
+                        "SMOTEN": im.SMOTEN}[sampler]
+        super().__init__(parameters=parameters,
+                         component_obj=None,
+                         random_seed=random_seed)
+
+    def fit(self, X, y):
+        """Fits the Oversampler to the data.
+
+        Arguments:
+            X (ww.DataFrame): Training features
+            y (ww.DataColumn): Target features
+
+        Returns:
+            self
+        """
+        super().fit(X, y)
+        self._initialize_oversampler(X, y, self.sampler)
+
+    def _initialize_oversampler(self, X, y, sampler_class):
+        """Initializes the oversampler with the given sampler_ratio or sampler_ratio_dict. If a sampler_ratio_dict is provided, we will opt to use that.
+        Otherwise, we use will create the sampler_ratio_dict dictionary.
+
+        Arguments:
+            X (ww.DataFrame): Training features
+            y (ww.DataColumn): Target features
+            sampler_class (imblearn.BaseSampler): The sampler we want to initialize
+        """
+        _, _, _, y_pd = self._prepare_data(X, y)
+        sampler_params = {k: v for k, v in copy.copy(self.parameters).items() if k not in ['sampling_ratio', 'sampling_ratio_dict']}
+        if self.parameters['sampling_ratio_dict'] is not None and len(self.parameters['sampling_ratio_dict']):
```

Comment:
What happens when a user passes in a `sampling_ratio_dict` but the keys don't match the actual labels in the target?

My concern is that by the time the target makes it to the sampler in AutoMLSearch, the target has been encoded. But users would need to know about the encoding ahead of time to specify this correctly. I don't think that's ideal. Users are much more likely to specify the dictionary using the unencoded labels because that's the data they see prior to running search.

Reply:
Ah yeah, that's a good point. I didn't think about this.

I think the best way to solve this would be to remove support for passing in a dictionary. This means that sampling for every target will be purely based on the float value passed for `sampling_ratio`, which will allow us to avoid issues of the dictionary keys differing from the targets we have currently

The con with this is that this removes functionality to finely alter the number of samples per each class. I can't think of a way that would allow us to avoid the issue and allow the dictionary input, since the target encoding is done in the pipeline class. What do you think @freddyaboulton? Do you have ideas for this?

Reply:
Thanks for following up @bchen1116 ! 

I think there's a third option, which is to leave as-is and not allow the use of the dictionary in `AutoMLSearch`. I think I prefer that for now? We still give users fine-tuned control of how much oversampling is done per class but they just have to do it prior to search. I think that's ok.

I think the solution you lay out in 2 is technically feasible but I don't like that the pipeline has to modify the parameters of one of its components mid-way during evaluation.

Reply:
Mmm I see. That'd mean that if users define pipelines outside of AutoMLSearch, we should allow them to run into potentially confusing errors? ie when they have string target values and they pass a dictionary to the pipeline, we would then say those keys don't exist since we do encode them.

I think if we go with this approach, it'd be nice to file an issue to figure out how to best resolve this issue

Reply:
Ah, good point! Now that I think about it, I don't know if my proposal made a ton of sense at all. I was thinking that if users really wanted fine-grained control, they would create an oversampling component with a dict, oversample the data, and then pass the oversampled data to AutoMLSearch but that's not ideal because the sampling is not part of the pipeline.

Ok, so taking a step back, the problem is that users will likely misconfigure the keys of the sampling dictionary because they don't know about how the label encoding happens. 

With that in mind, I agree that our two options are those that you laid out before! Maybe a third one is that we only allow integer-valued keys ranging from 0 to # classes - 1. In essence we require users to pass in the encoded target values as keys. This may not be too harsh, the encoder encodes in ascending order so I think its straightforward for users to adapt the keys.

Given all this, I think we should go with option 1 (not support the dict) or option 3. I'm still hesitant on option 2 but maybe my opinion will change once I see the implementation. One thing to keep in mind is that we're probably going to represent the label encoder as a component soon. Maybe this is a point in favor of passing state around like @dsherry proposed in #2110! (but I still vote against writing separate encoder/decoder components 😂  )